contracts/AdExRecoveryDAO.sol

```diff
@@ -0,0 +1,157 @@
+        delete proposers[proposer];
+        emit LogRemoveProposer(proposer, now);
+    }
+
+    /**
+     * @notice Only proposers are allowed to a propose recovery request
+     * @param request The details of the recovery request
+     */
+    function proposeRecovery(RecoveryRequestLibrary.RecoveryRequest memory request) external {
+        require(proposers[msg.sender] == true, 'ONLY_WHITELISTED_PROPOSER');
+        bytes32 recoveryId = request.hash();
+        recovery[recoveryId] = recoveryDelay.add(now);
+        emit LogProposeRecovery(recoveryId, msg.sender, request.identity, now);
+    }
+    
+    /**
+     * @notice Any proposer can finalize a propose recovery request after the delay period
+     * @param request The details of the recovery request
+     */
+    function finalizeRecovery(RecoveryRequestLibrary.RecoveryRequest memory request) external {
+        bytes32 recoveryId = request.hash();
+        require(proposers[msg.sender] == true, 'ONLY_WHITELISTED_PROPOSERS');
+        require(recovery[recoveryId] > 0 , 'RECOVERY_REQUEST_DOES_NOT_EXIST');
+        require(now >= recovery[recoveryId], 'ACTIVE_DELAY');
+        
+        Identity.Transaction[] memory recoverTransaction = new Identity.Transaction[](1);
+        recoverTransaction[0] = Identity.Transaction(
+            request.identity,
+            Identity(request.identity).nonce(),
+            address(0),
+            0,
+            request.identity,
+            0,
+            abi.encodeWithSelector(
+                Identity.setAddrPrivilege.selector,
+                request.newWalletAddress,
+                uint8(Identity.PrivilegeLevel.Transactions)
+            )
+        );
+
+        Identity(request.identity).executeBySender(recoverTransaction);
+        delete recovery[recoveryId];
+        emit LogFinalizeRecovery(recoveryId, msg.sender, request.identity, now);
+    }
+
+    /**
+     * @notice Admin, any proposer or the identity being recovered can cancel a recovery request
+     * @param request The details of the recovery request
+     */
+    function cancelRecovery(RecoveryRequestLibrary.RecoveryRequest memory request) external {
+        require(
+            request.identity == msg.sender ||
+            msg.sender == adminAddr || 
+            proposers[msg.sender] == true, 
+            'ONLY_IDENTITY_PROPOSER_OR_ADMIN_CAN_CANCEL'
+        );
+        bytes32 recoveryId = request.hash();
+        require(recovery[recoveryId] > 0, 'RECOVERY_REQUEST_DOES_NOT_EXIST');
+        delete recovery[recoveryId];
+        emit LogCancelRecovery(recoveryId, msg.sender, now);
+    }
+    
+    /**
+     * @notice Only Admin can replace themself
+     * @param newAdmin The address of the new admin
+     */
+    function changeAdmin(address newAdmin) external {
+        require(msg.sender == adminAddr, 'ONLY_ADMIN_CAN_CALL');
+        require(newAdmin != adminAddr, 'INVALID_NEW_ADMIN');
```

Comment:
maybe `require(admin != address(0), 'INVALID_ADMIN')` or it is on purpose without that check.